gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/SpinnakerUserInfoTokenServices.groovy

```diff
@@ -0,0 +1,128 @@
+import org.springframework.security.web.authentication.preauth.PreAuthenticatedAuthenticationToken
+import org.springframework.stereotype.Component
+
+/**
+ * ResourceServerTokenServices is an interface used to manage access tokens. The UserInfoTokenService object is an
+ * implementation of that interface that uses an access token to get the logged in user's data (such as email or
+ * profile). We want to customize the Authentication object that is returned to include our custom (Kork) User.
+ */
+class SpinnakerUserInfoTokenServices implements ResourceServerTokenServices {
+  @Autowired
+  ResourceServerProperties sso
+
+  @Autowired
+  UserInfoTokenServices userInfoTokenServices
+
+  @Autowired
+  CredentialsService credentialsService
+
+  @Autowired
+  UserRolesProvider userRolesProvider
+
+  @Autowired
+  UserInfoMapping userInfoMapping
+
+  @Value('${spring.oauth2.client.userAuthorizationUri}')
+  URL userAuthorizationUri
+
+  @Override
+  OAuth2Authentication loadAuthentication(String accessToken) throws AuthenticationException, InvalidTokenException {
+    OAuth2Authentication oAuth2Authentication = userInfoTokenServices.loadAuthentication(accessToken)
+
+    Map details = oAuth2Authentication.userAuthentication.details as Map
+
+    if (!isUserFromRestrictedDomain(details)) {
+      throw new BadCredentialsException("User's email does not match the expected domain.")
+    }
+
+    def username = details[userInfoMapping.username] as String
+    def roles = userRolesProvider.loadRoles(username)
+
+    User spinnakerUser = new User(
+        email: details[userInfoMapping.email] as String,
+        firstName: details[userInfoMapping.firstName] as String,
+        lastName: details[userInfoMapping.lastName] as String,
+        allowedAccounts: credentialsService.getAccountNames(roles),
+        roles: roles,
+        username: username).asImmutable()
+
+    PreAuthenticatedAuthenticationToken authentication = new PreAuthenticatedAuthenticationToken(
+        spinnakerUser,
+        null /* credentials */,
+        spinnakerUser.authorities
+    )
+
+    // impl copied from UserInfoTokenServices
+    OAuth2Request storedRequest = new OAuth2Request(null, sso.clientId, null, true /*approved*/,
+                                                    null, null, null, null, null);
+
+    return new OAuth2Authentication(storedRequest, authentication)
+  }
+
+  @Override
+  OAuth2AccessToken readAccessToken(String accessToken) {
+    return userInfoTokenServices.readAccessToken(accessToken)
+  }
+
+  boolean isUserFromRestrictedDomain(Map details) {
+    if (!userInfoMapping.domainRestriction) {
+      // Nothing to look for in the user info response, therefore all users allowed.
+      return true
+    }
+
+    HttpUrl userAuthUri = HttpUrl.parse(userAuthorizationUri.toString())
+    String restrictedDomain = userAuthUri.queryParameter(userInfoMapping.domainRestriction)
```

Comment:
Travis and I just talked through this, I think he's going to try and make this more general. Something like "in the response Map, this key should have this value, otherwise fail".